python/cuml/manifold/t_sne.pyx

```diff
@@ -0,0 +1,472 @@
+        y : array-like (device or host) shape = (n_samples, 1)
+                y contains a label per row.
+                Acceptable formats: cuDF Series, NumPy ndarray, Numba device
+                ndarray, cuda array interface compliant array like CuPy
+        """
+        cdef int n, p
+        cdef cumlHandle* handle_ = <cumlHandle*><size_t>self.handle.getHandle()
+        assert(handle_ != NULL)
+
+        if len(X.shape) != 2:
+            raise ValueError("data should be two dimensional")
+
+        cdef uintptr_t X_ptr
+        if self._should_downcast:
+            _X, X_ptr, n, p, dtype = to_cuda(X, order='C',
+                                             convert_to_dtype=np.float32)
+        else:
+            _X, X_ptr, n, p, dtype = to_cuda(X, order='C',
+                                             check_dtype=np.float32)
+
+        if n <= 1:
+            raise ValueError("There needs to be more than 1 sample to build "
+                             "nearest the neighbors graph")
+
+        self.n_neighbors = min(n, self.n_neighbors)
+        if self.perplexity > n:
+            print("[Warn] Perplexity = {} should be less than the "
+                  "# of datapoints = {}.".format(self.perplexity, n))
+            self.perplexity = n
+
+        # Prepare output embeddings
+        Y = cuda.device_array(
+            (n, self.n_components),
+            order="F",
+            dtype=np.float32)
+
+        cdef uintptr_t embed_ptr = Y.device_ctypes_pointer.value
+
+        # Find best params if learning rate method is adaptive
+        if self.learning_rate_method=='adaptive' and self.method=="barnes_hut":
+            if self.verbose:
+                print("Learning rate is adpative. In TSNE paper, "
+                      "it has been shown that as n->inf, "
+                      "Barnes Hut works well if n_neighbors->30, "
+                      "learning_rate->20000, early_exaggeration->24.")
+                print("cuML uses an adpative method."
+                      "n_neighbors decreases to 30 as n->inf. "
+                      "Likewise for the other params.")
+            if n <= 2000:
+                self.n_neighbors = min(max(self.n_neighbors, 90), n)
+            else:
+                # A linear trend from (n=2000, neigh=100) to (n=60000,neigh=30)
+                self.n_neighbors = max(int(102 - 0.0012 * n), 30)
+            self.pre_learning_rate = max(n / 3.0, 1)
+            self.post_learning_rate = self.pre_learning_rate
+            self.early_exaggeration = 24.0 if n > 10000 else 12.0
+            if self.verbose:
+                print("New n_neighbors = {}, "
+                      "learning_rate = {}, "
+                      "exaggeration = {}".format(self.n_neighbors,
+                                                 self.pre_learning_rate,
+                                                 self.early_exaggeration))
+
+        assert(<void*> X_ptr != NULL and <void*> embed_ptr != NULL)
```

Comment:
Don’t use assert, raise exceptions. Also this is not needed here if I’m not mistaken

Reply:
Oh yes seems like I love asserting lol